features/config/service/api/src/main/java/org/opennms/features/config/service/util/PropertiesCacheWithCm.java

```diff
@@ -0,0 +1,168 @@
+    private final ConfigurationManagerService cm;
+    private final ConfigUpdateInfo configKey;
+    private final Lock lock = new ReentrantLock();
+    private boolean needReload = false;
+
+    PropertiesHolder(final ConfigurationManagerService cm, ConfigUpdateInfo configKey) {
+      this.cm = cm;
+      this.configKey = configKey;
+      boolean shouldListenToConfigChanges = Boolean.getBoolean(CHECK_LAST_MODIFY_STRING);
+      if(shouldListenToConfigChanges) {
+        // we want to be notified when config has changed
+        cm.registerReloadConsumer(configKey, (ConfigUpdateInfo key) -> this.needReload = true);
+      }
+      properties = null;
+    }
+
+    private Optional<Map<String, Object>> read() {
+      Optional<Map<String, Object>> result;
+      try {
+        result = this.cm.getJSONStrConfiguration(configKey.getConfigName(), configKey.getConfigId())
+                .map(PropertiesConversionUtil::jsonToMap);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      needReload = false;
+      return result;
+    }
+
+    private void write() throws IOException {
+      Map<String, Object> entries = new HashMap<>();
+      for(Entry<?, ?> entry : this.properties.entrySet()) {
+        entries.put(entry.getKey().toString(), entry.getValue());
+      }
+      this.cm.updateConfiguration(this.configKey.getConfigName(), this.configKey.getConfigId(), mapToJsonString(entries));
+    }
+
+    public Map<String, Object> get() throws IOException {
+      lock.lock();
+      try {
+        if (properties == null || needReload) {
+          this.properties = read().orElse(new ConcurrentHashMap<>());
+        }
+        return properties;
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    public void setProperty(final String key, final Object value) throws IOException {
+      lock.lock();
+      try {
+        // first we do get to make sure the properties are loaded
+        get();
+        if (!value.equals(get().get(key))) {
+          get().put(key, value);
+          write();
+        }
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    public Object getProperty(final String key) throws IOException {
+      lock.lock();
+      try {
+        return get().get(key);
+      } finally {
+        lock.unlock();
+      }
+
+    }
+  }
+
+  protected final Cache<ConfigUpdateInfo, PropertiesHolder> m_cache;
+
+  private final ConfigurationManagerService cm;
+
+  public PropertiesCacheWithCm(final ConfigurationManagerService cm) {
+    this(CacheBuilder.newBuilder(), cm);
+  }
+
+
+  protected PropertiesCacheWithCm(final CacheBuilder<Object, Object> cacheBuilder, final ConfigurationManagerService cm) {
+    m_cache = cacheBuilder
+        .expireAfterAccess(SystemProperties.getInteger(CACHE_TIMEOUT, DEFAULT_CACHE_TIMEOUT), TimeUnit.SECONDS)
+        .build();
+    this.cm = cm;
+  }
+
+  private PropertiesHolder getHolder(final ConfigUpdateInfo configKey) throws IOException {
+
```

Comment:
Just a small suggestion, rename it as configName. So it can align with XMLs. It is a little bit confuse between configKey & key.

Reply:
fixed. I renamed it to configIdentifier since it contains the configName and configKey